rpcperms/middleware_handler.go

```diff
@@ -0,0 +1,448 @@
+package rpcperms
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/golang/protobuf/proto"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/macaroons"
+)
+
+var (
+	// errShuttingDown is the error that's returned when the server is
+	// shutting down and a request cannot be served anymore.
+	errShuttingDown = errors.New("server shutting down")
+
+	// errTimeoutReached is the error that's returned if any of the
+	// middleware's tasks is not completed in the given time.
+	errTimeoutReached = errors.New("intercept timeout reached")
+
+	// errClientQuit is the error that's returned if the client closes the
+	// middleware communication stream before a request was fully handled.
+	errClientQuit = errors.New("interceptor RPC client quit")
+)
+
+// MiddlewareHandler is a type that communicates with a middleware over the
+// established bi-directional RPC stream. It sends messages to the middleware
+// whenever the custom business logic implemented there should give feedback to
+// a request or response that's happening on the main gRPC server.
+type MiddlewareHandler struct {
+	// lastRequestID is the ID of the last request that was forwarded to the
+	// middleware.
+	//
+	// NOTE: Must be used atomically!
+	lastRequestID uint64
+
+	middlewareName string
+
+	customCaveatName string
+
+	receive func() (*lnrpc.RPCMiddlewareResponse, error)
+
+	send func(request *lnrpc.RPCMiddlewareRequest) error
+
+	interceptRequests chan *interceptRequest
+
+	timeout time.Duration
+
+	// params are our current chain params.
+	params *chaincfg.Params
+
+	// done is closed when the rpc client terminates.
+	done chan struct{}
+
+	// quit is closed when lnd is shutting down.
+	quit chan struct{}
+
+	wg sync.WaitGroup
+}
+
+// NewMiddlewareHandler creates a new handler for the middleware with the given
+// name and custom caveat name.
+func NewMiddlewareHandler(name, customCaveatName string,
+	receive func() (*lnrpc.RPCMiddlewareResponse, error),
+	send func(request *lnrpc.RPCMiddlewareRequest) error,
+	timeout time.Duration, params *chaincfg.Params,
+	quit chan struct{}) *MiddlewareHandler {
+
+	// We explicitly want to log this as a warning since intercepting any
```

Comment:
I assume users would need to activate this config _manually_ and also obtain a special (possibly with additional locked down perms on disk) macaroon in order to interact w/ the service?

Reply:
Yeah, the middleware needs a macaroon with the `macaroon:write` permission to interact with the service to register itself. There currently is no global config to turn the entire feature on or off. Do you think we should add a flag? And should the feature be on or off by default?

Reply:
> Do you think we should add a flag? And should the feature be on or off by default?

Makes sense to have it behind a flag IMO, the feature is super powerful, so I think we want to observe it a bit in the wild to see how we can make it more secure and error prone before was make it a new default feature.